kernel/reduction.ml

```diff
@@ -915,16 +917,31 @@ let gen_conv cv_pb ?(l2r=false) ?(reds=TransparentState.full) env ?(evars=defaul
   in
     if b then ()
     else
-      let _ = clos_gen_conv reds cv_pb l2r evars env univs (univs, checked_universes) t1 t2 in
+      let _ = clos_gen_conv reds cv_pb l2r evars env univs (univs, checked_universes) (inject t1) (inject t2) in
         ()
 
 let conv = gen_conv CONV
 let conv_leq = gen_conv CUMUL
 
+
+let gen_conv_fconstr cv_pb ?(l2r=false) ?(reds=TransparentState.full) env ?(evars=default_evar_handler) t1 t2 =
+  let univs = Environ.universes env in
+  let b =
+    if cv_pb = CUMUL then leq_constr_univs univs t1 (term_of_fconstr t2)
```

Comment:
Isn't this `term_of_fconstr` defeating the purpose?

Reply:
If I understand correctly, it is not *completely* defeating the purpose. My understanding is that this is doing a syntactic equality check (which I assume is potentially faster than the conversion check). The actual conversion check will still use the original `fconstr` which has (potentially) useful reduction information already inside of it.

I am experimenting with removing the conversion entirely.

Reply:
Here's a non-scientific performance comparison:

| version | foo | foo2 |
|------------|-----|-------|
| master | 0.266s | 0.001s |
| with `term_of_fconstr` | 0.214s | 0.001s |
| no `term_of_fconstr` | 0.189s | 0.001s |

Here's the test code.

```coq
Fixpoint get_unit (n : nat) : Type :=
  match n with
  | 0 => unit
  | S n => get_unit n
  end.

Fixpoint make (n : nat) : Type :=
  match n with
  | 0 => unit
  | S n => get_unit n -> make n
  end.

Fixpoint make2 (n : nat) : Type :=
  match n with
  | 0 => unit
  | S n => unit -> make2 n
  end.

Definition foo : make 1000.
  compute. intros. apply tt.
Qed.

Definition foo2 : make2 1000.
  compute. intros. apply tt.
Qed.

Lemma xxx : unit.
Proof.
  apply foo; exact tt.
Time Qed.

Lemma xxx2 : unit.
Proof.
  apply foo2; exact tt.
Time Qed.
```